File: src/Applications/Orbital.Web/Orbital.Forecast.BlazorApplication/BlazorApp.Client/Library/DynamicLinkLibraries/Events/Event.Portable/Arrows/EventLink.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Runtime.Serialization;


using CategoryTheory;
using Event.Interfaces;

namespace Event.Portable.Arrows
{
    /// <summary>
    /// Link between event and its handler
    /// </summary>
    public class EventLink : CategoryArrow,  IRemovableObject
    {
        #region Fields

        IEvent target;

        IEventHandler source;

        #endregion

        #region Ctor

        /// <summary>
        /// Default constructor
        /// </summary>
        public EventLink()
        {
        }

  

        #endregion

        #region Overriden Members

        /// <summary>
        /// The source of this arrow
        /// </summary>
        public override ICategoryObject Source
        {
            get
            {
                return source as ICategoryObject;
            }
            set
            {
                source = value.GetSource<IEventHandler>();
            }
        }

        /// <summary>
        /// The target of this arrow
        /// </summary>
        public override ICategoryObject Target
        {
            get
            {
                return target as ICategoryObject;
            }
            set
            {
                target = value.GetTarget<IEvent>();
                source.Add(target);
            }
        }

        #endregion

        #region IRemovableObject Members

        void IRemovableObject.RemoveObject()
        {
            if ((source == null) | (target == null))
            {
                return;
            }
            source.Remove(target);
            source = null;
        }


        #endregion

    

    }
}

File: src/Web/reactprojectsimplegame/src/Library/Event/Objects/EventLink.ts
import { CategoryArrow } from "../../CategoryArrow";
import { OwnNotImplemented } from "../../ErrorHandler/OwnNotImplemented";
import type { ICategoryObject } from "../../Interfaces/ICategoryObject";
import type { IDesktop } from "../../Interfaces/IDesktop";
import type { IEvent } from "../../Interfaces/IEvent";
import type { IEventHandler } from "../../Interfaces/IEventHandler";

export class EventLink extends CategoryArrow
{
    constructor(desktop: IDesktop, name: string) {
        super(desktop, name)
    }

    setSource(source: ICategoryObject): void {
        var h = this.getObjectT<IEventHandler, ICategoryObject>(source, "IEventHandler")
        if (h.length == 0) {
            throw new OwnNotImplemented("EVENT LINK");
        }
        this.handler = h[0]
        super.setSource(source)
    }

    setTarget(target: ICategoryObject): void {
        this.target = target;
        var e = this.getObjectT<IEvent, ICategoryObject>(target, "IEvent")
        if (e.length == 0) {
            throw new OwnNotImplemented("Event link");
        }
        this.event = e[0]
        this.handler.addEventToHandler(e[0])
        super.setTarget(target)
 }

    event !: IEvent

    handler !: IEventHandler 
    /*

         public IEventHandler Source { get => source; }

        /// <summary>
        /// The event (target)
        /// </summary>
        public IEvent Target { get => target; }

    */

}